src/Data/MIME.hs

```diff
@@ -776,20 +847,48 @@ mime'
   -> BodyHandler MIME
 mime' takeTillEnd h = RequiredBody $ case view contentType h of
   ct | view ctType ct == "multipart" ->
-    case preview (rawParameter "boundary") ct of
-      Nothing -> FailedParse MultipartBoundaryNotSpecified <$> takeTillEnd
-      Just boundary ->
-        (Multipart <$> multipart takeTillEnd boundary)
-        <|> (FailedParse MultipartParseFail <$> takeTillEnd)
+        case prepMultipart ct of
+          Left err              -> FailedParse err <$> takeTillEnd
+          Right (sub, boundary) ->
+            Multipart sub boundary <$> multipart takeTillEnd boundary
+            <|> FailedParse MultipartParseFail <$> takeTillEnd
      | matchContentType "message" (Just "rfc822") ct ->
         (Encapsulated <$> message (mime' takeTillEnd))
         <|> (FailedParse EncapsulatedMessageParseFail <$> takeTillEnd)
-  _ -> part
+  _ -> Part <$> takeTillEnd
   where
-    part = Part <$> takeTillEnd
+    prepMultipart ct =
+      (,) <$> parseSubtype ct <*> parseBoundary ct
+    parseBoundary ct =
+      getRequiredParam "boundary" ct
+      >>= over _Left (InvalidParameterValue "boundary") . makeBoundary
```

Comment:
Am I correct that `makeBoundary` is used for boundary parsing, not for a caller to create a new boundary? For that you will need to use `genBoundary`?

Reply:
library users can `makeBoundary` if they want, but I think `genBoundary` would be more convenient for most use cases.